src/main/java/io/github/thebusybiscuit/slimefun4/core/services/MetricsService.java

```diff
@@ -0,0 +1,234 @@
+            .enableCookieManagement(false)
+            .cookieSpec("ignoreCookies");
+    }
+
+    public MetricsService(SlimefunPlugin plugin) {
+        this.plugin = plugin;
+        this.metricFile = new File(plugin.getDataFolder(), REPO_NAME + ".jar");
+    }
+
+    /**
+     * This method loads the metric module and starts the metrics collection.
+     */
+    public void start() {
+        boolean newlyDownloaded = false;
+        if (!metricFile.exists()) {
+            info(REPO_NAME + " does not exist, downloading...");
+            if (!download(getLatestVersion())) {
+                plugin.getLogger().warning("Failed to start metrics as the file could not be downloaded.");
+                return;
+            }
+            newlyDownloaded = true;
+        }
+
+        try {
+            // Load the jar file into a child class loader using the SF PluginClassLoader
+            // as a parent.
+            moduleClassLoader = URLClassLoader.newInstance(new URL[] { metricFile.toURI().toURL() },
+                plugin.getClass().getClassLoader());
+            Class<?> cl = moduleClassLoader.loadClass("dev.walshy.sfmetrics.MetricsModule");
+
+            metricVersion = cl.getPackage().getImplementationVersion();
+
+            // If it has not been newly downloaded, auto-updates are on AND there's a new version
+            // then cleanup, download and start
+            if (!newlyDownloaded
+                && hasAutoUpdates()
+                && checkForUpdate(metricVersion)
+            ) {
+                info("Cleaning up and re-loading Metrics.");
+                cleanUp();
+                start();
+                return;
+            }
+
+            // Finally, we're good to start this.
+            Method start = cl.getDeclaredMethod("start");
+            String s = cl.getPackage().getImplementationVersion();
+
+            // This is required to be sync due to bStats.
+            Slimefun.runSync(() -> {
+                try {
+                    start.invoke(null);
+                    info("Metrics build " + s + " started.");
+                } catch (Exception e) {
+                    plugin.getLogger().log(Level.WARNING, "Failed to start metrics.", e);
+                }
+            });
+        } catch (Exception e) {
+            plugin.getLogger().log(Level.WARNING,
+                "Failed to load the metrics module. Maybe the jar is corrupt?", e);
+        }
+    }
+
+    /**
+     * This will close the child classloader and mark all the resources held under this no longer
+     * in use, they will be cleaned up the next GC run.
+     */
+    public void cleanUp() {
+        try {
+            if (this.moduleClassLoader != null)
+                this.moduleClassLoader.close();
+        } catch (IOException e) {
+            plugin.getLogger().log(Level.WARNING,
+                "Could not clean up module class loader. Some memory may have been leaked.");
+        }
+    }
```

Comment:
Is this method even needed here? Or is it for when someone uses /reload?

Reply:
It's to cleanup the class loader, if we didn't close it and a reload happened (or an update to the module) then we may leak classes and these class resources.

This is ran when re-starting (on new version) and onDisable